File: src/routes/reports.ts
import express from "express";
import { Client } from 'pg';
import crypto from 'crypto';
import fs from 'fs';

export default (app: express.Application, database: Client, websockets: Map<string, Map<string, WebSocket[]>>) => {
    app.get('/reports', (req: express.Request, res: express.Response) => {
        database.query(`SELECT * FROM reports`, async (err, dbRes) => {
            if (!err) {
                const reports = dbRes.rows.filter(x => x.school === res.locals.school);
                database.query(`SELECT * FROM users`, async (err, dbRes) => {
                    if (!err) {
                        res.send(reports.map(x => {
                            x.file = JSON.parse(x.file);
                            x.author = { id: x.author, name: dbRes.rows.find(y => y?.id === x.author)?.name ?? "Deleted user" };
                            x.date = Number(x.date);
                            delete x.school;
                            return x;
                        }));
                    } else {
                        console.log(err);
                        res.status(500).send({ error: "Server error." });
                    }
                });
            } else {
                console.log(err);
                res.status(500).send({ error: "Server error." });
            }
        });
    });

    app.post('/reports', (req: express.Request, res: express.Response) => {
        const files = fs.readdirSync(__dirname + '/../../files');
        const report = {
            id: crypto.randomUUID(),
            title: req.body.title,
            file: req.body.file ?? [],
            author: res.locals.user,
            date: Date.now(),
            school: res.locals.school
        };
        database.query(`SELECT * FROM users`, async (err, dbRes) => {
            if (!err) {
                if (JSON.parse(dbRes.rows.find(x => x.id === res.locals.user).administrator).includes(res.locals.school)) {
                    if (report.title && report.file && files.includes(report.file.id)) {
                        const users = dbRes.rows;
                        database.query(`INSERT INTO reports (id, title, file, author, date, school) VALUES($1, $2, $3, $4, $5, $6)`, [report.id, report.title, JSON.stringify(report.file), report.author, report.date.toString(), report.school], (err, dbRes) => {
                            if (!err) {
                                let websocketiedReport = { ...report };
                                websocketiedReport.author = { id: websocketiedReport.author, name: users.find(y => y?.id === websocketiedReport.author)?.name ?? "Deleted user" };
                                users.map(x => x.id).forEach(receiver => {
                                    websockets.get(res.locals.school)?.get(receiver)?.forEach(websocket => {
                                        websocket.send(JSON.stringify({ event: 'newReport', ...websocketiedReport }));
                                    });
                                });
                                res.status(201).send({});
                            } else {
                                console.log(err);
                                res.status(500).send({ error: "Server error." });
                            }
                        });
                    } else {
                        res.status(400).send({ error: "Missing required argument." });
                    }
                } else {
                    res.status(403).send({ error: "Not authorized." });
                }
            } else {
                console.log(err);
                res.status(500).send({ error: "Server error." });
            }
        });
    });

    app.patch('/reports/*', (req: express.Request, res: express.Response) => {
        const urlParamsValues: string[] = Object.values(req.params);
        const reportId = urlParamsValues
            .map((x) => x.replace(/\//g, ''))
            .filter((x) => {
                return x !== '';
            })[0];

        database.query(`SELECT * FROM reports`, async (err, dbRes) => {
            if (!err) {
                const oldReport = dbRes.rows.find(x => x.id === reportId);
                if (oldReport) {
                    database.query(`SELECT * FROM users`, async (err, dbRes) => {
                        if (!err) {
                            if (JSON.parse(dbRes.rows.find(x => x.id === res.locals.user).administrator).includes(res.locals.school)) {
                                if (req.body.title) {
                                    const users = dbRes.rows;
                                    database.query(`UPDATE reports SET title = $1 WHERE id = $2`, [req.body.title, reportId], (err, dbRes) => {
                                        if (!err) {
                                            users.map(x => x.id).forEach(receiver => {
                                                websockets.get(res.locals.school)?.get(receiver)?.forEach(websocket => {
                                                    websocket.send(JSON.stringify({ event: 'editedReport', id: reportId, newTitle: req.body.title }));
                                                });
                                            });
                                            res.send({});
                                        } else {
                                            console.log(err);
                                            res.status(500).send({ error: "Server error." });
                                        }
                                    });
                                } else {
                                    res.status(400).send({ error: "Missing required argument." });
                                }
                            } else {
                                res.status(403).send({ error: "Not authorized." });
                            }
                        } else {
                            console.log(err);
                            res.status(500).send({ error: "Server error." });
                        }
                    });
                } else {
                    res.status(404).send({ error: "Not found." });
                }
            } else {
                console.log(err);
                res.status(500).send({ error: "Server error." });
            }
        });
    });

    app.delete('/reports/*', (req: express.Request, res: express.Response) => {
        const urlParamsValues: string[] = Object.values(req.params);
        const reportId = urlParamsValues
            .map((x) => x.replace(/\//g, ''))
            .filter((x) => {
                return x !== '';
            })[0];

        database.query(`SELECT * FROM users`, async (err, dbRes) => {
            if (!err) {
                if (JSON.parse(dbRes.rows.find(x => x.id === res.locals.user).administrator).includes(res.locals.school)) {
                    const users = dbRes.rows;
                    database.query(`DELETE FROM reports WHERE id = $1`, [reportId], async (err, dbRes) => {
                        if (!err) {
                            users.map(x => x.id).forEach(receiver => {
                                websockets.get(res.locals.school)?.get(receiver)?.forEach(websocket => {
                                    websocket.send(JSON.stringify({ event: 'deletedReport', id: reportId }));
                                });
                            });
                            res.send({});
                        } else {
                            console.log(err);
                            res.status(500).send({ error: "Server error." });
                        }
                    });
                } else {
                    res.status(403).send({ error: "Not authorized." });
                }
            } else {
                console.log(err);
                res.status(500).send({ error: "Server error." });
            }
        });
    });
};

File: routes/reports.go
package routes

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/Scunea/Backend/types"
	"github.com/Scunea/Backend/utils"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
)

func StartReports(r *gin.Engine, db *sqlx.DB) {
	r.GET("/reports", AuthRequired(db, false), func(c *gin.Context) {
		var schoolId = c.GetHeader("school")
		var reports = utils.GetReports(db, schoolId)

		var reportsParsed = utils.ParseReports(db, reports)

		c.JSON(http.StatusOK, reportsParsed)
	})

	r.POST("/reports", AuthRequired(db, false), func(c *gin.Context) {
		var body types.ReportPostRequest
		c.BindJSON(&body)

		var schoolId = c.GetHeader("school")
		var user = utils.GetMyUser(db, c)

		if utils.Type(db, user, schoolId) == "administrator" {
			if len(body.Title) > 0 && len(body.File.Id) > 0 {

				if utils.FilesExist([]types.IdAndName{body.File}) {

					var report = types.ReportParsed{
						Id:    uuid.NewString(),
						Title: body.Title,
						File:  body.File,
						Author: types.IdAndName{
							Id:   user.Id,
							Name: user.Name,
						},
						Date:   time.Now().UnixMilli(),
						School: schoolId,
					}

					_, err := db.Exec("INSERT INTO reports (id, title, file, author, date, school) VALUES ($1, $2, $3, $4, $5, $6)", report.Id, report.Title, utils.UnParse(report.File), report.Author.Id, fmt.Sprint(report.Date), report.School)
					if err == nil {
						c.JSON(http.StatusOK, gin.H{
							"success": "Report sent.",
						})

						var users = utils.GetUsersFromSchool(db, schoolId)

						var newReport = types.NewReport{
							Event:  "newReport",
							Report: report,
						}

						for _, user := range users {
							utils.SendNotification(db, user.Id, newReport)
						}

						utils.SendWebsocket(schoolId, "", newReport)
					} else {
						c.JSON(http.StatusBadRequest, gin.H{
							"error": "Error sending report.",
						})
					}
				} else {
					c.JSON(http.StatusNotFound, gin.H{
						"error": "File not found.",
					})
				}
			} else {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Title or File is empty.",
				})
			}
		} else {
			c.JSON(http.StatusUnauthorized, gin.H{
				"error": "You are not an administrator.",
			})
		}
	})

	r.PATCH("/reports/:id", AuthRequired(db, false), func(c *gin.Context) {
		var reportId = c.Param("id")

		var body types.ReportPatchRequest
		c.BindJSON(&body)

		var schoolId = c.GetHeader("school")
		var user = utils.GetMyUser(db, c)

		var report, ok = utils.GetReport(db, reportId, schoolId)

		if ok {
			if utils.Type(db, user, schoolId) == "administrator" {
				var file types.IdAndName

				json.Unmarshal([]byte(report.File), &file)

				date, err := strconv.ParseInt(report.Date, 0, 64)

				if err == nil {

					var report = types.ReportParsed{
						Id:    report.Id,
						Title: body.Title,
						File:  file,
						Author: types.IdAndName{
							Id:   user.Id,
							Name: user.Name,
						},
						Date:   date,
						School: schoolId,
					}

					_, err := db.Exec("UPDATE reports SET title = $1 WHERE id = $2", report.Title, report.Id)
					if err == nil {
						c.JSON(http.StatusOK, gin.H{
							"success": "Report edited.",
						})
						utils.SendWebsocket(schoolId, "", types.EditedReport{
							Event:  "editedReport",
							Id:     report.Id,
							Report: report,
						})
					} else {
						c.JSON(http.StatusBadRequest, gin.H{
							"error": "Error editing report.",
						})
					}
				} else {
					c.JSON(http.StatusBadRequest, gin.H{
						"error": "Error editing report.",
					})
				}
			} else {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "You are not the author of this report.",
				})
			}
		} else {
			c.JSON(http.StatusNotFound, gin.H{
				"error": "Report not found.",
			})
		}
	})

	r.DELETE("/reports/:id", AuthRequired(db, false), func(c *gin.Context) {
		var reportId = c.Param("id")

		var schoolId = c.GetHeader("school")
		var user = utils.GetMyUser(db, c)

		var _, ok = utils.GetReport(db, reportId, schoolId)

		if ok {
			if utils.Type(db, user, schoolId) == "administrator" {
				_, err := db.Exec("DELETE FROM reports WHERE id = $1", reportId)
				if err == nil {
					c.JSON(http.StatusOK, gin.H{
						"success": "Report deleted.",
					})
					utils.SendWebsocket(schoolId, "", types.DeletedReport{
						Event: "deletedReport",
						Id:    reportId,
					})
				} else {
					c.JSON(http.StatusBadRequest, gin.H{
						"error": "Error deleting report.",
					})
				}
			} else {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "You are not an administrator.",
				})
			}
		} else {
			c.JSON(http.StatusNotFound, gin.H{
				"error": "Report not found.",
			})
		}
	})
}
